src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -475,15 +392,31 @@ private void ValidateImplementedMethod(MethodInfo methodInfoBody, MethodInfo met
         {
             if ((methodInfoBody.IsVirtual || methodInfoBody.IsStatic) &&
                 (methodInfoDeclaration.IsAbstract || methodInfoDeclaration.IsVirtual) &&
-                methodInfoBody.ReturnType.Equals(methodInfoDeclaration.ReturnType))
+                methodInfoDeclaration.ReturnType.IsAssignableFrom(methodInfoBody.ReturnType))
             {
                 ParameterInfo[] bodyParameters = methodInfoBody.GetParameters();
                 ParameterInfo[] declarationParameters = methodInfoDeclaration.GetParameters();
                 if (bodyParameters.Length == declarationParameters.Length)
                 {
                     for (int i = 0; i < bodyParameters.Length; i++)
                     {
-                        if (!bodyParameters[i].ParameterType.Equals(declarationParameters[i].ParameterType))
+                        Type? bodyType = bodyParameters[i].ParameterType;
+                        Type? declType = declarationParameters[i].ParameterType;
+
+                        if (bodyType.IsArray != declType.IsArray ||
+                            bodyType.IsByRef != declType.IsByRef ||
+                            bodyType.IsPointer != declType.IsPointer)
+                        {
+                            throw ArgumentExceptionInvalidMethodOverride(methodInfoDeclaration.Name);
+                        }
+
+                        if (bodyType.HasElementType || declType.HasElementType)
```

Comment:
Is the `if` check necessary?

Reply:
Yes, do you mean call GetElementType() even HasElementType was false?